Implement VirtIO block device

otteryc opened the pull request:
This commit migrates @shengwen-tw's brilliant virtio implementation from semu, with the following modification:

1. Rename virtio_blk_reg_read/write to virtio_blk_read/write The original virtio_blk_read verified whether the virtio MMIO was aligned to 4. Since rv32emu does not enforce alignment checks for UART and PLIC devices or raise misalignment exceptions, this verification has been omitted.

2. Implement MMIO_VIRTIOBLK

3. Implement vblk_new() and vblk_delete() 
These functions align with the conventions used for UART and PLIC devices.

4. Introduce new argument '-x vblk:<disk_image>' for virtio-blk disk image

This pull request has been tested on:
1. Linux 6.8.0 @ Intel core i7-7700K
2. Mac OS Sequoia 15.1 @ Apple Silicon M2

To try out virtio-blk:
Generate ext4 image file for virtio block device in Unix-like system:
```shell
$ dd if=/dev/zero of=disk.img bs=4M count=32
$ mkfs.ext4 disk.img
```
Boot the Guest OS
```shell
$ ./build/rv32emu -k <kernel_img_path> -i <rootfs_img_path> -x vblk:disk.img
```
Mount the virtual block device and create a test file after booting:
```shell
$ mkdir mnt
$ mount /dev/vda mnt
$ echo "rv32emu" > mnt/emu.txt
$ umount mnt
```
Reboot the system and re-mount the virtual block device, the written file should remain existing.

otteryc committed:
Migrate virtio-blk from semu

This commit migrates @shengwen-tw's brilliant virtio implementation
from semu, with the following modification:

1. Rename virtio_blk_reg_read/write to virtio_blk_read/write
The original virtio_blk_read verified whether the virtio MMIO was
aligned to 4. Since rv32emu does not enforce alignment checks for UART
and PLIC devices or raise misalignment exceptions, this verification
has been omitted.

2. Implement MMIO_VIRTIOBLK

3. Implement vblk_new() and vblk_delete()
These functions align with the conventions used for UART and PLIC
devices.

4. Introduce new argument '-x vblk:<disk_image>' for virtio-blk disk
image

5. Fix the size of addr in virtq_desc structure.

Co-authored-by: Shengwen Cheng <shengwen1997.tw@gmail.com>

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: fd051c24c9c7143cb43fed6de603139aa116326e | Previous: caef4ad8f25bf2162a8f2fb6475d081f1608083d | Ratio |
|-|-|-|-|
| `Dhrystone` | `1299` Average DMIPS over 10 runs | `1336` Average DMIPS over 10 runs | `1.03` |
| `Coremark` | `959.33` Average iterations/sec over 10 runs | `974.943` Average iterations/sec over 10 runs | `1.02` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

jserv commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,437 @@
+#include <assert.h>
```
Add copyright notice:
```
/*
 * rv32emu is freely redistributable under the MIT License. See the file
 * "LICENSE" for information on usage and redistribution of this file.
 */
```

See https://github.com/sysprog21/rv32emu/blob/master/src/riscv.c

jserv commented on src/system.c:
```diff
@@ -28,6 +29,15 @@ void emu_update_uart_interrupts(riscv_t *rv)
     plic_update_interrupts(attr->plic);
 }
 
+void emu_update_vblk_interrupts(riscv_t *rv)
```
Declare internal functions as `static`.

otteryc replied:
Should a new PR be created to make `emu_update_uart_interrupts` static? Since UART is not within the scope of this PR, would it be acceptable to declare both `emu_update_uart_interrupts` and `emu_update_vblk_interrupts` as static in a new commit within this PR instead?

jserv replied:
> Should a new PR be created to make `emu_update_uart_interrupts` static? 

You can create another pull request for such purpose. After merging that, rebase.

jserv reviewed (changes requested):
1. Specify Shengwen Cheng as co-author in git commit messages. See https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors
2. Consolidate GitHub Actions to automate virtio-blk test procedure.

jserv commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,442 @@
+    uint16_t new_avail = ram[queue->QueueAvail] >> 16;
+    if (new_avail - queue->last_avail > (uint16_t) queue->QueueNum)
+        return (fprintf(stderr, "size check fail\n"),
+                virtio_blk_set_fail(vblk));
```
Split into two statements. It would be better for static analyzer to locate the source lines.

jserv commented on src/main.c:
```diff
@@ -73,6 +74,7 @@ static void print_usage(const char *filename)
 #if RV32_HAS(SYSTEM) && !RV32_HAS(ELF_LOADER)
             "  -k <image> : use <image> as kernel image\n"
             "  -i <image> : use <image> as rootfs\n"
+            "  -v <image> : use <image> as virtio-blk disk image\n"
```
Avoid the potentially confusing `-v` option. Instead, use `-x` as the "extended" option, with syntax `-x vblk:<image>` for specifying the virtio-blk image. When virtio-net is ready to merge, use `-x vnet:tap` to utilize TAP/TUN as the virtio-net backend.

vacantron commented:
Does it have any instructions / documentations for demonstrating or testing of this newly added feature?

ChinYikMing commented:
rv32emu use `snake_case` naming scheme (check [CONTRIBUTING.md](https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md)), but I see [CamelCase](https://en.wikipedia.org/wiki/Camel_case) in the implementation. @otteryc Can you unify the naming scheme or any strong reason to use CamelCase?

otteryc commented:
@vacantron @ChinYikMing Thank you for the insightful comments.
CamelCase names have been replaced with snake_case, and a new section has been added to the README to demonstrate virtio-blk.

otteryc commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,446 @@
+
+static inline uint32_t vblk_preprocess(virtio_blk_state_t *vblk, uint32_t addr)
+{
+#if RV32_HAS(SYSTEM) && !RV32_HAS(ELF_LOADER)
```
This is because `MEM_SIZE` is not defined in the Makefile when both features are enabled, leading to a compilation error.
I am wondering if there is a better solution for this.

ChinYikMing replied:
> This is because `MEM_SIZE` is not defined in the Makefile when both features are enabled, leading to a compilation error. I am wondering if there is a better solution for this.

When both of them are not enabled, the default memory size would be `FULL4G`.

I do not get compilation error by `make system ENABLE_SYSTEM=1` on Linux machine using gcc 9.4.0. Could you elaborate more?

Edit: gcc 9.4.0

otteryc replied:
```diff
@@ -80,10 +80,8 @@ static void virtio_blk_set_fail(virtio_blk_state_t *vblk)
 static inline uint32_t vblk_preprocess(virtio_blk_state_t *vblk UNUSED,
                                        uint32_t addr)
 {
-#if RV32_HAS(SYSTEM) && !RV32_HAS(ELF_LOADER)
     if ((addr >= MEM_SIZE) || (addr & 0b11))
         return virtio_blk_set_fail(vblk), 0;
-#endif
```
```shell 
$ make ENABLE_ELF_LOADER=1 ENABLE_SYSTEM=1
  CC	build/devices/virtio-blk.o
src/devices/virtio-blk.c: In function ‘vblk_preprocess’:
src/devices/virtio-blk.c:83:18: error: ‘MEM_SIZE’ undeclared (first use in this function)
   83 |     if ((addr >= MEM_SIZE) || (addr & 0b11))
      |                  ^~~~~~~~
src/devices/virtio-blk.c:83:18: note: each undeclared identifier is reported only once for each function it appears in
make: *** [mk/system.mk:26: build/devices/virtio-blk.o] Error 1
```

This behavior is identical with [this ci job](https://github.com/sysprog21/rv32emu/actions/runs/12923763274/job/36045942745#step:8:328)

ChinYikMing replied:
Okay, I can reproduce it now.

```diff
@@ -80,10 +80,8 @@ static void virtio_blk_set_fail(virtio_blk_state_t *vblk)
 static inline uint32_t vblk_preprocess(virtio_blk_state_t *vblk UNUSED,
                                        uint32_t addr)
 {
-#if RV32_HAS(SYSTEM) && !RV32_HAS(ELF_LOADER)
     if ((addr >= MEM_SIZE) || (addr & 0b11))
         return virtio_blk_set_fail(vblk), 0;
-#endif
```

Please try this by modifying the upper level of Makefile:
```diff
$(call set-feature, FULL4G)
ifeq ($(call has, FULL4G), 1)
-$(OUT)/main.o: CFLAGS += -DMEM_SIZE=0xFFFFFFFFULL # 2^{32} - 1
+CFLAGS += -DMEM_SIZE=0xFFFFFFFFULL # 2^{32} - 1
endif
```

ChinYikMing commented on src/devices/virtio.h:
```diff
@@ -0,0 +1,110 @@
+
+#define VIRTIO_VENDOR_ID 0x12345678
+
+#define VIRTIO_STATUS__DRIVER_OK 4
```
Why double underscore in the middle?

otteryc replied:
Though it is not explicitly mentioned, those macros with double underscore are defined by virtio specification according to my observation.

ChinYikMing replied:
I found `DRIVER_OK` in [here](https://docs.oasis-open.org/virtio/virtio/v1.3/virtio-v1.3.html) but do not see double underscore. Can you show me where you see it?

otteryc replied:
I originally thought it was used to distinguish feature bits and status bits from others.
However, since the difference is not trivial, I am removing the double underscores.
Thanks!

ChinYikMing commented on src/devices/virtio.h:
```diff
@@ -0,0 +1,110 @@
+#define VIRTIO_VENDOR_ID 0x12345678
+
+#define VIRTIO_STATUS__DRIVER_OK 4
+#define VIRTIO_STATUS__DEVICE_NEEDS_RESET 64
```
Ditto.

ChinYikMing commented on src/devices/virtio.h:
```diff
@@ -0,0 +1,110 @@
+#define VIRTIO_STATUS__DRIVER_OK 4
+#define VIRTIO_STATUS__DEVICE_NEEDS_RESET 64
+
+#define VIRTIO_INT__USED_RING 1
```
Ditto.

ChinYikMing commented on src/devices/virtio.h:
```diff
@@ -0,0 +1,110 @@
+#define VIRTIO_STATUS__DEVICE_NEEDS_RESET 64
+
+#define VIRTIO_INT__USED_RING 1
+#define VIRTIO_INT__CONF_CHANGE 2
```
Ditto.

ChinYikMing commented on src/main.c:
```diff
@@ -56,6 +56,7 @@ static char *prof_out_file;
 static char *opt_kernel_img;
 static char *opt_rootfs_img;
 static char *opt_bootargs;
+static char *opt_virtio_img;
```
Rename to `opt_virtio_blk_img`.

ChinYikMing commented on src/devices/virtio.h:
```diff
@@ -0,0 +1,110 @@
+    uint16_t next;
+};
+
+#define IRQ_VBLK 3
```
Rename to `IRQ_VBLK_SHIFT` to consistent with `IRQ_UART_SHIFT` in 'src/devices/uart.h'.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,446 @@
+#define VBLK_QUEUE_NUM_MAX 1024
+#define VBLK_QUEUE (vblk->queues[vblk->queue_sel])
+
+#define PRIV(x) ((struct virtio_blk_config *) x->priv)
```
Can you rename `PRIV(x)` to something else like `VBLK_PRIV(x)` to avoid confusion with `PRIV(rv)` for accessing RISC-V instance attributes?

Alternatively, for better abstraction and to facilitate access to the `priv` field by other Virtio components if neccessary, consider refining the macro to `VIRTIO_PRIV(x, type) ((type *) x->priv)`.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,446 @@
+        fprintf(stderr,
+                "Exceeded the number of virtio-blk devices that can be "
+                "allocated.\n");
+        exit(2);
```
Use `exit(EXIT_FAILURE)`.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,446 @@
+    int disk_fd = open(disk_file, O_RDWR);
+    if (disk_fd < 0) {
+        fprintf(stderr, "could not open %s\n", disk_file);
+        exit(2);
```
Ditto.

ChinYikMing reviewed (commented):
Change the description of this pull request, adding information for others to build the system emulator and test the virtio-blk feature. This will make it easier for others to access the information without needing to refer to the README but also keep them in README :) .

jserv commented:
Compilation warnings reported by GCC:
```
src/devices/virtio-blk.c: In function ‘vblk_preprocess’:
src/devices/virtio-blk.c:80:60: error: unused parameter ‘vblk’ [-Werror=unused-parameter]
   80 | static inline uint[32](https://github.com/sysprog21/rv32emu/actions/runs/12930766138/job/36065876824?pr=539#step:9:33)_t vblk_preprocess(virtio_blk_state_t *vblk, uint32_t addr)
      |                                        ~~~~~~~~~~~~~~~~~~~~^~~~
```

Simply add `UNUSED` (see `src/common.h`) to get rid of the above.

ChinYikMing commented on README.md:
````diff
@@ -77,9 +77,24 @@ $ make ENABLE_SYSTEM=1 system
+$ dd if=/dev/zero of=disk.img bs=4M count=32
+$ mkfs.ext4 disk.img
+```
+Mount the virtual block device and create a test file after booting:
````
Aware the user that the mount/umount operation needs root permission.

otteryc replied:
In busybox, users login with root priviliges by default.
Do you mean to make users aware of mounting the image outside the emulation?

jserv replied:
Quote from semu:
> For demonstration purposes, ext4 is used for file system mounting. ext4 is a native Linux filesystem, offering stability, high capacity, reliability, and performance while requiring minimal maintenance. The mkfs.ext4 command can create an ext4 file system from disk partitions. This command is a symbolic link of the [mke2fs](https://man7.org/linux/man-pages/man8/mke2fs.8.html) command, and its usage is the same as the mke2fs command.

No superuser is needed.

ChinYikMing replied:
Buildroot supports creating user via option: `BR2_ROOTFS_USERS_TABLES`. I have created a non-root user test1.

Login using user test1:
```
Welcome to Buildroot
buildroot login: test1
Password:
$ mkdir mnt
$ mount /dev/vda mnt
mount: you must be root
```

The mount is not available to user test1. Need to grant root permission via `su` command.

After granting the root permission for user test1 via `su` command:
```
$ su
$ mount /dev/vda mnt
[   59.215841] EXT4-fs (vda): mounted filesystem with ordered data mode. Quota mode: disabled.
$ echo "rv32emu" > mnt/emu.txt
$  umount mnt
[   73.701061] EXT4-fs (vda): unmounting filesystem.
```

Thus, aware the root permission might be needed is necessary since the login user might not be root.

jserv commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,447 @@
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "common.h"
```
You don't have to include `common.h`. The top-level `Makefile` already specifies `CFLAGS += -include src/common.h -Isrc/`.

RinHizakura commented on src/main.c:
```diff
@@ -118,6 +120,13 @@ static bool parse_args(int argc, char **args)
             emu_argc++;
             break;
+        case 'x':
+            if (!strncmp("vblk:", optarg, 5))
```
Why must we introduce such complexity with the `vblk:` prefix? The brilliant way is to use just the disk file name here and consider adding the long option for different block devices if required.

jserv replied:
> Why must we introduce such complexity with the `vblk:` prefix?

VirtIO features like disk, network, and sound can be dynamically enabled using the `-x` option, inspired by [Oracle/Sun Java virtual machine](https://www.oracle.com/java/technologies/javase/vmoptions-jsp.html). For example: `-x vblk:<disk> -x vnet:tap0 -x vsnd:alsa`. The long command options might be confusing if we want to specify more VirtIO devices, such as a secondary virtio-blk.

RinHizakura replied:
Okay. Thanks

RinHizakura commented on src/riscv.h:
```diff
@@ -571,6 +573,10 @@ typedef struct {
+
+    /* virtio-blk device */
+    uint32_t *disk;
+    virtio_blk_state_t *vblk;
```
Ok it seems like UART and VIRTIO_BLK come from semu, and PLIC may have a little bit different design here......

I am just thinking of wrapping these things in `virtio_blk_t`, but maybe we can refactor these in other PR if required. This is not a critical issue.

ChinYikMing commented on README.md:
```diff
@@ -77,9 +77,24 @@ $ make ENABLE_SYSTEM=1 system
+$ mkdir mnt
+$ mount /dev/vda mnt
+$ echo "rv32emu" > mnt/emu.txt
+$ umount
```
Should be `umount mnt`? If yes, edit the desciption of the pull request as well.

otteryc replied:
Indeed, thanks!

ChinYikMing replied:
Please also update the description of the PR. Thanks!

ChinYikMing commented on README.md:
````diff
@@ -77,9 +77,24 @@ $ make ENABLE_SYSTEM=1 system
 ```shell
 $ make ENABLE_SYSTEM=1
-$ build/rv32emu -k <kernel_img_path> -i <rootfs_img_path>
+$ build/rv32emu -k <kernel_img_path> -i <rootfs_img_path> [-x vblk:<virtio_img_path>]
````
Should be `virtio_blk_img_path`.

ChinYikMing commented on Makefile:
```diff
@@ -189,7 +189,7 @@ endif
 $(call set-feature, FULL4G)
 ifeq ($(call has, FULL4G), 1)
-$(OUT)/main.o: CFLAGS += -DMEM_SIZE=0xFFFFFFFFULL # 2^{32} - 1
+	CFLAGS += -DMEM_SIZE=0xFFFFFFFFULL # 2^{32} - 1
```
Remove the `\t` since it is not a recipe.

ChinYikMing commented on .ci/virtio-blk.sh:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
```
Can you reuse the [script](https://github.com/sysprog21/rv32emu/blob/master/.ci/boot-linux.sh) of boot Linux instead of creating a new one? By adjusting the script to accept a CLI parameter that determines whether to test the virtio-blk feature.

jserv commented on .ci/virtio-blk.sh:
```diff
@@ -0,0 +1,54 @@
+RUN_LINUX="build/rv32emu ${OPTS}"
+
+dd if=/dev/zero of=build/disk.img bs=4M count=32
+mkfs.ext4 build/disk.img
```
Check the presence of `mkfs.ext4`. That is, something like https://github.com/sysprog21/semu/blob/master/Makefile#L20-L32

otteryc committed:
CI: Add virtio-blk verification

Virtio-blk verification has been integrated into boot-linux test, and
is enabled by default.
The p7zip-full package, providing command '7z', is added to
dependencies in order to validate the success of writes to the
Virtio-blk device.

This commit also removes -b argument when starting the emulator, as it
is refined to be used to accept custom bootargs.

ChinYikMing commented on README.md:
````diff
@@ -77,9 +77,24 @@ $ make ENABLE_SYSTEM=1 system
+# echo "rv32emu" > mnt/emu.txt
+# umount mnt
+```
+Reboot and re-mount the system, the written file should remain existing.
````
Proofread as: "Reboot and re-mount the virtual block device, the written file should remain existing."

jserv commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,62 @@ function ASSERT {
+ENABLE_VBLK=1
+type dd >/dev/null 2>&1 || ENABLE_VBLK=0
+(type mkfs.ext4 >/dev/null 2>&1 || type $(brew --prefix e2fsprogs)/sbin/mkfs.ext4) >/dev/null 2>&1 || ENABLE_VBLK=0
+type 7z >/dev/null 2>&1 || ENABLE_VBLK=0
```
Why is 7z required?

otteryc replied:
I believe using `7z` to list files in `disk.img` is the simplest and most efficient method to verify whether the file `emu.txt` written by the guest OS exists.
Other approaches have limitations:
- Mounting `disk.img` requires root privileges, which may not always be feasible.
- Inspecting the image with `debugfs` provides the desired result but involves navigating a complex interactive interface.
- `dumpe2fs` fail with an error, "Wrong magic number for **Ext2** Image Header while trying to open disk.img" when attempting to process `disk.img`.

jserv replied:
Can we simply use `cpio` or `ar` instead?

otteryc replied:
> Can we simply use `cpio` or `ar` instead?

Neither `cpio`, `ar` nor `zcat` recognizes the image file, and after a quick review of their man pages, I couldn’t find any useful options to address this issue.

jserv replied:
We can verify the success of filesystem creation by checking the return code from `mkfs.ext4` directly, rather than using additional tools like 7z. Since 7z is not included in standard Linux distributions, avoiding such external dependencies makes the testing process more accessible to other developers and simplifies collaboration.

otteryc replied:
> We can verify the success of filesystem creation by checking the return code from `mkfs.ext4` directly, rather than using additional tools like 7z. Since 7z is not included in standard Linux distributions, avoiding such external dependencies makes the testing process more accessible to other developers and simplifies collaboration.

I completely agree that avoiding extra tools like `7z` would significantly improve portability and make the testing process more accessible. However, checkin the return code of `mkfs.ext4` merely verifies that the filesystem was created, but it does not confirm that Virtio-blk is functioning correctly.
For example:
```shell
$ dd if=/dev/zero of=disk.img bs=4M count=32
32+0 records in
32+0 records out
134217728 bytes (134 MB, 128 MiB) copied, 0.314109 s, 427 MB/s
$ mkfs.ext4 disk.img
mke2fs 1.47.0 (5-Feb-2023)
Discarding device blocks: done
Creating filesystem with 32768 4k blocks and 32768 inodes

Allocating group tables: done
Writing inode tables: done
Creating journal (4096 blocks): done
Writing superblocks and filesystem accounting information: done

$ mkfs.ext4 disk.img
mke2fs 1.47.0 (5-Feb-2023)
disk.img contains a ext4 file system
	created on Mon Jan 27 23:22:21 2025
Proceed anyway? (y,N) n

$ echo $?
1
```
As shown above, the second call to mkfs.ext4 returns 1 because the filesystem was already created by the first call, even though the emulator was never started.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,443 @@
+    case _(QueueNumMax):
+        return VBLK_QUEUE_NUM_MAX;
+    case _(QueueReady):
+        return VBLK_QUEUE.ready ? 1 : 0;
```
Rewrite as: `return (uint32_t) VBLK_QUEUE.ready` since `VBLK_QUEUE.ready` is a boolean that already evaluates to either true or false.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,443 @@
+
+#define DISK_BLK_SIZE 512
+
+#define VBLK_DEV_CNT_MAX 1
```
Hmm, does this refer to the virtio block limit? I believe it can attach multiple virtual block devices, can't it?

otteryc replied:
> Hmm, does this refer to the virtio block limit? I believe it can attach multiple virtual block devices, can't it?

Certainly, multiple virtual block devices can be attached. However, after a quick review, it appears that each `vblk` device requires an independent virtqueue at least. Additionally, unlike QEMU, rv32emu’s current implementation uses a hardcoded device tree at build time. This could lead to significant complications with MMIO register mapping and negatively impact performance due to the extra polling required.

If possible, I would prefer to leave this for future extension, when the device tree and MMIO mapping can be adjusted at runtime.

jserv replied:
>  I would prefer to leave this for future extension, when the device tree and MMIO mapping can be adjusted at runtime.

Add corresponding FIXME/TODO comments.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,443 @@
+#define _(reg) VIRTIO_##reg
+    switch (addr) {
+    case _(MagicValue):
+        return 0x74726976;
```
Define these magic value to a semantic name, just like `VIRTIO_VENDOR_ID` indicates the vendor ID.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,443 @@
+    case _(MagicValue):
+        return 0x74726976;
+    case _(Version):
+        return 2;
```
Ditto.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,443 @@
+    case _(Version):
+        return 2;
+    case _(DeviceID):
+        return 2;
```
Ditto.

ChinYikMing commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,443 @@
+    case _(Status):
+        return vblk->status;
+    case _(ConfigGeneration):
+        return 0;
```
Ditto.

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,62 @@ function ASSERT {
 OPTS+=" -b build/minimal.dtb "
+if [ "$ENABLE_VBLK" -eq "1" ]; then
+    dd if=/dev/zero of=build/disk.img bs=4M count=32
+    mkfs.ext4 build/disk.img || $(brew --prefix e2fsprogs)/sbin/mkfs.ext4 build/disk.img
```
No macOS CI is integrated yet. Is it suitable to include brew utility since we do not actually use it for now?

otteryc replied:
This was added based on [this review](https://github.com/sysprog21/rv32emu/pull/539#discussion_r1928168618). While we do not currently utilize the brew utility, its inclusion is harmless and imposes minimal runtime overhead.

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,62 @@ function ASSERT {
 
+ENABLE_VBLK=1
+type dd >/dev/null 2>&1 || ENABLE_VBLK=0
+(type mkfs.ext4 >/dev/null 2>&1 || type $(brew --prefix e2fsprogs)/sbin/mkfs.ext4) >/dev/null 2>&1 || ENABLE_VBLK=0
```
Ditto.

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,62 @@ function ASSERT {
-printf "\n[ ${COLOR_G}${MESSAGES[$ret]}${COLOR_N} ]\n"
+printf "\nBoot Linux Test: [ ${COLOR_G}${MESSAGES[$ret]}${COLOR_N} ]\n"
+if [ "$ENABLE_VBLK" -eq "1" ]; then 
+    file_list=`7z l build/disk.img`
```
Rewrite as `file_list`= $(7z l build/disk.img)

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,62 @@ function ASSERT {
 TIMEOUT=50
 OPTS=" -k build/linux-image/Image "
 OPTS+=" -i build/linux-image/rootfs.cpio "
 OPTS+=" -b build/minimal.dtb "
```
I believe this can be remove since `-b` has been refined to be used to accept custom bootargs. Cc. @RinHizakura

otteryc replied:
Should it be removed within this pull request?

ChinYikMing replied:
> Should it be removed within this pull request?

Yes, go ahead. Please also mention the rationale in the commit message.

shengwen-tw commented on src/devices/virtio.h:
```diff
@@ -0,0 +1,110 @@
+};
+
+struct virtq_desc {
+    uint32_t addr;
```
According to [Virtual I/O Device (VIRTIO) Version 1.3](https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html#x1-430005)

*2.7.5 The Virtqueue Descriptor Table*

```C
struct virtq_desc { 
        /* Address (guest-physical). */ 
        le64 addr; 
        /* Length. */ 
        le32 len; 
        /* The flags as indicated above. */ 
        le16 flags; 
        /* We chain unused descriptors via this, too */ 
        le16 next; 
}; 
```

We mistakenly defined `virtq_desc.addr` as 32 bits.
Please consider correcting it here and submitting a pull request for semu.

jserv commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,62 @@ function ASSERT {
 cleanup
 
+ENABLE_VBLK=1
+type dd >/dev/null 2>&1 || ENABLE_VBLK=0
```
Use `which` for better shell compatibility. The command `which` even exits in busybox, widely used in embedded Linux.

otteryc replied:
> Use `which` for better shell compatibility. The command `which` even exits in busybox, widely used in embedded Linux.

The use of `type` has been replaced by `which`, but I personally have some question about this change.

`type` is defined in [IEEE 1003.1 (POSIX)](https://ieeexplore.ieee.org/document/10555529) (Pages 3473-3475) and is therefore guaranteed to be available in all POSIX-compliant shells. On the other hand, `which` appears to be implementation-defined and not standardized by POSIX.

jserv replied:
> `type` is defined in [IEEE 1003.1 (POSIX)](https://ieeexplore.ieee.org/document/10555529) (Pages 3473-3475) and is therefore guaranteed to be available in all POSIX-compliant shells. On the other hand, `which` appears to be implementation-defined and not standardized by POSIX.

That being said, the use of "which" command is a de facto standard and is widely adopted by most Linux and BSD variants. The POSIX standard, on the other hand, cannot be fully compliant with by most systems.

jserv reviewed (changes requested):
Rebase the latest master branch to resolve Arm64 host breakage on CI.

jserv commented on src/devices/virtio-blk.c:
```diff
@@ -0,0 +1,444 @@
+                                       uint32_t addr)
+{
+    if ((addr >= MEM_SIZE) || (addr & 0b11))
+        return virtio_blk_set_fail(vblk), 0;
```
Don't do this. Ensure that only one function call appears in return statement.

otteryc commented:
**HELP WANTED**

I found that the current implementation is incorrect when `HAVE_MMAP` is set to `0`. While testing, I encountered a compiler error due to an undeclared function in lines 53 and 80 related to `getpagesize()`. This function is declared in unistd.h in Unix-like systems, but `unistd.h` is not included when `HAVE_MMAP` is 0.

I am currently not in Tainan and do not have access to an MSVC environment. Therefore, I am unsure whether this is a bug or if `getpagesize()` is properly defined in other MSVC headers. Any insights or verification would be greatly appreciated!

jserv commented:
> I am currently not in Tainan and do not have access to an MSVC environment. Therefore, I am unsure whether this is a bug or if `getpagesize()` is properly defined in other MSVC headers. Any insights or verification would be greatly appreciated!

MSVC support can be disregarded as it is not functional. While initial testing was performed on Windows 8 during the my original code snapshot in 2020, there has never been a working Windows build of rv32emu. Of course, if you are interested in a Windows port, pull requests are always welcome. At present, concentrate on the condition `HAVE_MMAP=1`.

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,61 @@ function ASSERT {
+if [ "$ENABLE_VBLK" -eq "1" ]; then
+    dd if=/dev/zero of=build/disk.img bs=4M count=32
+    mkfs.ext4 build/disk.img || $(brew --prefix e2fsprogs)/sbin/mkfs.ext4 build/disk.img
+    OPTS+=" -x vblk:build/disk.img "
```
`build/disk.img` is repeated in the changes. Replace them with a variable such as `VBLK_IMG=build/disk.img` and access it via `$(VBLK_IMG)`.

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,61 @@ function ASSERT {
+printf "\nBoot Linux Test: [ ${COLOR_G}${MESSAGES[$ret]}${COLOR_N} ]\n"
+if [ "$ENABLE_VBLK" -eq "1" ]; then 
+    file_list=$(7z l build/disk.img)
+    (echo $file_list | grep emu.txt >/dev/null 2>&1) || ret=4
```
Since `file_list` is only accessed here, is it better to streamline them into one line by removing `file_list` variable?

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,61 @@ function ASSERT {
+     "Fail to find emu.txt in $VBLK_IMG"\
 )
 
 COLOR_G='\e[32;01m' # Green
```
Introduce red color here such that: `COLOR_R='\e[31;01m' # Red`

ChinYikMing commented on .ci/boot-linux.sh:
```diff
@@ -16,31 +16,61 @@ function ASSERT {
+printf "\nBoot Linux Test: [ ${COLOR_G}${MESSAGES[$ret]}${COLOR_N} ]\n"
+if [ "$ENABLE_VBLK" -eq "1" ]; then 
+    7z l $VBLK_IMG | grep emu.txt >/dev/null 2>&1 || ret=4
+    printf "Virtio-blk Test: [ ${COLOR_G}${MESSAGES[$ret]}${COLOR_N} ]\n"
```
Show the VirtIO block error message (ret=4) in red instead of green since it is typically used for success.

otteryc replied:
> Show the VirtIO block error message (ret=4) in red instead of green since it is typically used for success.

Shall other error messages be changed into red as well?

ChinYikMing replied:
> > Show the VirtIO block error message (ret=4) in red instead of green since it is typically used for success.
> 
> Shall other error messages be changed into red as well?

Yes, let's make them consistent. Thanks!

otteryc commented:
Added TODO statement and rebased to the latest `master` branch.

jserv reviewed (changes requested):
Rebase the latest `master` branch and fix conflicts.

jserv commented:
Thank @otteryc for contributing!
